### ISSUE TITLE
IfcCurveSegment doesn't compile with IfcLengthMeasure for .Start or .Length

### ISSUE THREAD

**Issue by RickBrice** [IfcOpenShell/IfcOpenShell#3897]
I'm using the bSI validation service for this file https://github.com/RickBrice/IfcAlignmentDriver/blob/d5d2c837a0175d44a4bccc4e5d6cdb3f43450156/TestFiles/SpiralCurves/SC_S30W_Right_6b.ifc 

There are errors that IfcCurveSegment is using IfcNonNegativeLengthMeasure with negative values. The error is correct.
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/37087370/f48a5bf3-1d9c-48bf-a29d-72c529f6118f)

When I change the IFCOS C++ code that creates this file from

```
            auto curve_segment = new Schema::IfcCurveSegment(Schema::IfcTransitionCode::IfcTransitionCode_CONTSAMEGRADIENT, segment_curve_placement, 
               new Schema::IfcNonNegativeLengthMeasure(start_dist_along), 
               new Schema::IfcNonNegativeLengthMeasure(horizontal_length), 
               parent_curve);
```

to

```
            auto curve_segment = new Schema::IfcCurveSegment(Schema::IfcTransitionCode::IfcTransitionCode_CONTSAMEGRADIENT, segment_curve_placement, 
               new Schema::IfcLengthMeasure(start_dist_along), 
               new Schema::IfcLengthMeasure(horizontal_length), 
               parent_curve);
```
there are compile errors

```
1>F:\ARP\PGSuperIfcExtensions\IfcModelBuilder.cpp(579,61): error C2665: 'Ifc4x3_tc1::IfcCurveSegment::IfcCurveSegment': no overloaded function could convert all the argument types
1>F:\IfcOpenShell\src\ifcparse\Ifc4x3_tc1.h(20195,5): message : could be 'Ifc4x3_tc1::IfcCurveSegment::IfcCurveSegment(Ifc4x3_tc1::IfcTransitionCode::Value,Ifc4x3_tc1::IfcPlacement *,Ifc4x3_tc1::IfcCurveMeasureSelect *,Ifc4x3_tc1::IfcCurveMeasureSelect *,Ifc4x3_tc1::IfcCurve *)'
1>F:\ARP\PGSuperIfcExtensions\IfcModelBuilder.cpp(579,61): message : 'Ifc4x3_tc1::IfcCurveSegment::IfcCurveSegment(Ifc4x3_tc1::IfcTransitionCode::Value,Ifc4x3_tc1::IfcPlacement *,Ifc4x3_tc1::IfcCurveMeasureSelect *,Ifc4x3_tc1::IfcCurveMeasureSelect *,Ifc4x3_tc1::IfcCurve *)': cannot convert argument 3 from 'Ifc4x3_tc1::IfcLengthMeasure *' to 'Ifc4x3_tc1::IfcCurveMeasureSelect *'
```

because IfcLengthMeasure does not inherit from IfcCurveMeasureSelect.

Any recommendations for addressing this?

**Comment by aothms** [comment]
We have two options:

- Skip the validator for now and just use negative values inside a IfcNonNegativeLengthMeasure. Only the python-based validator checks the where rule on the numeric type. It does not inhibit using the negative number in C++.
- Upgrade to ADD2 where this is addressed. I will make sure the schema is available in v0.8.0. It's just a bit painful to do all this recompilation again.

**Comment by RickBrice** [comment]
Thanks - I'll stick with IfcNonNegativeLengthMeasure to keep the C++ development moving forward with the knowledge that eventually ADD2 will be supported and I'll have to revise the code to use IfcLengthMeasure at a later date.
